{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMAv+Xx4b0ggcn86JJg7GfU"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["!pip install torch_geometric\n","!pip install biopython\n","!pip install networkx"],"metadata":{"id":"t4vvwK7m6lNq","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1746577348310,"user_tz":240,"elapsed":33865,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"3694febd-ef1b-44f0-c275-d1b2d3103e0a"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting torch_geometric\n","  Downloading torch_geometric-2.6.1-py3-none-any.whl.metadata (63 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m63.1/63.1 kB\u001b[0m \u001b[31m1.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: aiohttp in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (3.11.15)\n","Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (2025.3.2)\n","Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (3.1.6)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (2.0.2)\n","Requirement already satisfied: psutil>=5.8.0 in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (5.9.5)\n","Requirement already satisfied: pyparsing in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (3.2.3)\n","Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (2.32.3)\n","Requirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from torch_geometric) (4.67.1)\n","Requirement already satisfied: aiohappyeyeballs>=2.3.0 in /usr/local/lib/python3.11/dist-packages (from aiohttp->torch_geometric) (2.6.1)\n","Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.11/dist-packages (from aiohttp->torch_geometric) (1.3.2)\n","Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.11/dist-packages (from aiohttp->torch_geometric) (25.3.0)\n","Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.11/dist-packages (from aiohttp->torch_geometric) (1.6.0)\n","Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.11/dist-packages (from aiohttp->torch_geometric) (6.4.3)\n","Requirement already satisfied: propcache>=0.2.0 in /usr/local/lib/python3.11/dist-packages (from aiohttp->torch_geometric) (0.3.1)\n","Requirement already satisfied: yarl<2.0,>=1.17.0 in /usr/local/lib/python3.11/dist-packages (from aiohttp->torch_geometric) (1.20.0)\n","Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch_geometric) (3.0.2)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->torch_geometric) (3.4.1)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->torch_geometric) (3.10)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->torch_geometric) (2.4.0)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->torch_geometric) (2025.4.26)\n","Downloading torch_geometric-2.6.1-py3-none-any.whl (1.1 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.1 MB\u001b[0m \u001b[31m10.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: torch_geometric\n","Successfully installed torch_geometric-2.6.1\n","Collecting biopython\n","  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (2.0.2)\n","Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m18.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: biopython\n","Successfully installed biopython-1.85\n","Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (3.4.2)\n"]}]},{"cell_type":"code","execution_count":2,"metadata":{"id":"GlS18Qnq5vdj","executionInfo":{"status":"ok","timestamp":1746577378019,"user_tz":240,"elapsed":27289,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}}},"outputs":[],"source":["import os\n","import torch\n","import torch.nn as nn\n","import torch.nn.functional as F\n","import pandas as pd\n","from torch_geometric.nn import GCNConv, global_mean_pool\n","from torch_geometric.data import Data, DataLoader\n","from sklearn.model_selection import train_test_split\n","from tqdm import tqdm\n","import requests\n","import networkx as nx\n","from scipy.spatial.distance import euclidean\n","from Bio.PDB import PDBParser, PDBList\n","from torch_geometric.nn import GCNConv, GraphNorm, global_mean_pool\n","\n","from torch.utils.data import Dataset"]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive', force_remount=True)\n","import pandas as pd\n","import numpy as np"],"metadata":{"id":"-DGZSZhI6VZC","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1746577395445,"user_tz":240,"elapsed":14427,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"7afd84dd-a99a-4bdf-b255-5f8ed9dfc3f9"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"code","source":["# df_fireprot = pd.read_csv('/content/drive/MyDrive/BMI_707_Project/707_Files_for_Colab/df_with_skipped_flag.csv')"],"metadata":{"id":"5NvdsdwS3RiI","executionInfo":{"status":"ok","timestamp":1746578140368,"user_tz":240,"elapsed":1354,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}}},"execution_count":29,"outputs":[]},{"cell_type":"code","source":["df_fireprot = pd.read_csv(\"/content/drive/MyDrive/707/data/fireprotdb_results.csv\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pEzd-stgIOhL","executionInfo":{"status":"ok","timestamp":1746581863328,"user_tz":240,"elapsed":888,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"72359a8f-e952-43b3-a3e2-8e0d8afa5f32"},"execution_count":76,"outputs":[{"output_type":"stream","name":"stderr","text":["<ipython-input-76-080b10c7c33b>:1: DtypeWarning: Columns (23,24,25,26,29) have mixed types. Specify dtype option on import or set low_memory=False.\n","  df_fireprot = pd.read_csv(\"/content/drive/MyDrive/707/data/fireprotdb_results.csv\")\n"]}]},{"cell_type":"code","source":["df_fireprot_w_ddG = df_fireprot[df_fireprot[\"ddG\"].notna()]\n","df_fireprot_w_ddG.shape"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"eZgfqRSNYqca","executionInfo":{"status":"ok","timestamp":1746582082889,"user_tz":240,"elapsed":42,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"78158e09-77b8-4e6b-e563-470e680db2a1"},"execution_count":83,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(39177, 35)"]},"metadata":{},"execution_count":83}]},{"cell_type":"code","source":["df_fireprot_no_dupl = df_fireprot_w_ddG.drop_duplicates()\n","df_fireprot_no_dupl.shape"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"urQnRzsHZk9U","executionInfo":{"status":"ok","timestamp":1746582087861,"user_tz":240,"elapsed":228,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"8b4aa1c8-c669-49f5-b519-922369bf447d"},"execution_count":85,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(12131, 35)"]},"metadata":{},"execution_count":85}]},{"cell_type":"code","source":["df_fireprot_w_pnames = df_fireprot_no_dupl[df_fireprot_no_dupl[\"protein_name\"].notna()]\n","df_fireprot_w_uniprot_id = df_fireprot_w_pnames[df_fireprot_w_pnames[\"uniprot_id\"].notna()]\n","df_fireprot_w_uniprot_id.shape"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"QMCsT_Sdab2i","executionInfo":{"status":"ok","timestamp":1746584645594,"user_tz":240,"elapsed":31,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"06811c6b-1be6-4bfc-e393-fe81d3dd3d2d"},"execution_count":141,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(12090, 35)"]},"metadata":{},"execution_count":141}]},{"cell_type":"code","source":["subset_cols = [\"protein_name\",\"chain\",\"position\", \"wild_type\", \"mutation\"]\n","df_groupby = df_fireprot_w_uniprot_id.groupby(subset_cols)\n","df_fire_prot_median_ddG = pd.DataFrame()\n","exp_ids = []\n","p_names = []\n","pbd_ids = []\n","chains = []\n","positions = []\n","wild_types = []\n","mutations = []\n","ddGs = []\n","sequences = []\n","is_in_cat_pock = []\n","is_ess = []\n","for name, group in tqdm(df_groupby):\n","  median_ddG_id = group[\"ddG\"].median()\n","  group[\"dist_to_median\"] = group[\"ddG\"].apply(lambda x: abs(x-median_ddG_id))\n","  group_sorted = group.sort_values(by=\"dist_to_median\")\n","  exp_ids.append(group_sorted.iloc[0][\"experiment_id\"])\n","  p_names.append(group_sorted.iloc[0][\"protein_name\"])\n","  pbd_ids.append(group_sorted.iloc[0][\"pdb_id\"])\n","  positions.append(group_sorted.iloc[0][\"position\"])\n","  wild_types.append(group_sorted.iloc[0][\"wild_type\"])\n","  mutations.append(group_sorted.iloc[0][\"mutation\"])\n","  ddGs.append(group_sorted.iloc[0][\"ddG\"])\n","  sequences.append(group_sorted.iloc[0][\"sequence\"])\n","  is_in_cat_pock.append(group_sorted.iloc[0][\"is_in_catalytic_pocket\"])\n","  is_ess.append(group_sorted.iloc[0][\"is_essential\"])\n","  chains.append(group_sorted.iloc[0][\"chain\"])\n","  # print(group_sorted)\n","\n","df_fire_prot_median_ddG = pd.DataFrame({\n","    \"experiment_id\": exp_ids,\n","    \"protein_name\": p_names,\n","    \"pdb_id\": pbd_ids,\n","    \"chain\": chains,\n","    \"position\": positions,\n","    \"wild_type\": wild_types,\n","    \"mutation\": mutations,\n","    \"ddG\": ddGs,\n","    \"sequence\": sequences,\n","    \"is_in_catalytic_pocket\": is_in_cat_pock,\n","    \"is_essential\": is_ess\n","})"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"T-RqivWfgVwv","executionInfo":{"status":"ok","timestamp":1746585303277,"user_tz":240,"elapsed":8998,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"9d364204-cf6a-4dbd-cb77-cb1baa195ff4"},"execution_count":156,"outputs":[{"output_type":"stream","name":"stderr","text":["100%|██████████| 5086/5086 [00:08<00:00, 572.25it/s]\n"]}]},{"cell_type":"code","source":["def mutate_sequence(row):\n","  sequence = list(row[\"sequence\"])\n","  # position = min(row[\"position\"]-1, len(sequence))\n","  sequence[row[\"position\"]-1] = row[\"mutation\"]\n","  return \"\".join(sequence)"],"metadata":{"id":"CJQnj5FfLCMy","executionInfo":{"status":"ok","timestamp":1746585314825,"user_tz":240,"elapsed":3,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}}},"execution_count":157,"outputs":[]},{"cell_type":"code","source":["df_fire_prot_median_ddG[\"mt_sequence\"] = df_fire_prot_median_ddG.apply(mutate_sequence, axis=1)"],"metadata":{"id":"aWY39IaUL6Vq","executionInfo":{"status":"ok","timestamp":1746586172173,"user_tz":240,"elapsed":155,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}}},"execution_count":160,"outputs":[]},{"cell_type":"code","source":["df_fire_prot_median_ddG[\"pdb_id_list\"] = df_fire_prot_median_ddG[\"pdb_id\"].apply(lambda x: list(set(x.split(\"|\"))))"],"metadata":{"id":"8ESeJIZjpYKl","executionInfo":{"status":"ok","timestamp":1746586236923,"user_tz":240,"elapsed":18,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}}},"execution_count":164,"outputs":[]},{"cell_type":"code","source":["df_fire_prot_median_ddG[\"pdb_id_len\"] = df_fire_prot_median_ddG[\"pdb_id_list\"].apply(lambda x: len(x))\n","df_fire_prot_median_ddG.sort_values(by=\"pdb_id_len\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":791},"id":"CTk7Jx6SppJ3","executionInfo":{"status":"ok","timestamp":1746586292080,"user_tz":240,"elapsed":356,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"5f4ed110-3627-4d61-c791-0211e0eb6086"},"execution_count":167,"outputs":[{"output_type":"execute_result","data":{"text/plain":["     experiment_id                          protein_name  \\\n","5081      PT008738  U1 small nuclear ribonucleoprotein A   \n","5080       VB01735  U1 small nuclear ribonucleoprotein A   \n","5079       VB04793  U1 small nuclear ribonucleoprotein A   \n","5078      PT008735  U1 small nuclear ribonucleoprotein A   \n","5077       VB01730  U1 small nuclear ribonucleoprotein A   \n","...            ...                                   ...   \n","898       PT000404               Dihydrofolate reductase   \n","966       PT014402               Dihydrofolate reductase   \n","965       PT002910               Dihydrofolate reductase   \n","964        VB05467               Dihydrofolate reductase   \n","963       PT002903               Dihydrofolate reductase   \n","\n","                        pdb_id chain  position wild_type mutation   ddG  \\\n","5081                      1OIA     A        86         Y        F  2.60   \n","5080                      1OIA     A        78         Y        F  0.00   \n","5079                      1OIA     A        78         Y        D  4.30   \n","5078                      1OIA     A        31         Y        S  2.30   \n","5077                      1OIA     A        31         Y        F  0.00   \n","...                        ...   ...       ...       ...      ...   ...   \n","898   1RX4|1DDR|1DDR|1DYJ|5DFR     A        31         F        V  1.50   \n","966   1RX4|1DDR|1DDR|1DYJ|5DFR     A       122         D        A  1.60   \n","965   1RX4|1DDR|1DDR|1DYJ|5DFR     A       121         G        Y  0.20   \n","964   1RX4|1DDR|1DDR|1DYJ|5DFR     A       121         G        V  1.10   \n","963   1RX4|1DDR|1DDR|1DYJ|5DFR     A       121         G        S -0.01   \n","\n","                                               sequence  \\\n","5081  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","5080  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","5079  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","5078  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","5077  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","...                                                 ...   \n","898   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","966   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","965   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","964   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","963   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","\n","      is_in_catalytic_pocket  is_essential  \\\n","5081                   False         False   \n","5080                   False         False   \n","5079                   False         False   \n","5078                   False         False   \n","5077                   False         False   \n","...                      ...           ...   \n","898                     True          True   \n","966                    False         False   \n","965                    False         False   \n","964                    False         False   \n","963                    False         False   \n","\n","                                            mt_sequence  \\\n","5081  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","5080  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","5079  MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...   \n","5078  MAVPETRPNHTIYINNLNEKIKKDELKKSLSAIFSQFGQILDILVS...   \n","5077  MAVPETRPNHTIYINNLNEKIKKDELKKSLFAIFSQFGQILDILVS...   \n","...                                                 ...   \n","898   MISLIAALAVDRVIGMENAMPWNLPADLAWVKRNTLNKPVIMGRHT...   \n","966   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","965   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","964   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","963   MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...   \n","\n","                   pdb_id_list  pdb_id_len  \n","5081                    [1OIA]           1  \n","5080                    [1OIA]           1  \n","5079                    [1OIA]           1  \n","5078                    [1OIA]           1  \n","5077                    [1OIA]           1  \n","...                        ...         ...  \n","898   [5DFR, 1DDR, 1DYJ, 1RX4]           4  \n","966   [5DFR, 1DDR, 1DYJ, 1RX4]           4  \n","965   [5DFR, 1DDR, 1DYJ, 1RX4]           4  \n","964   [5DFR, 1DDR, 1DYJ, 1RX4]           4  \n","963   [5DFR, 1DDR, 1DYJ, 1RX4]           4  \n","\n","[5086 rows x 14 columns]"],"text/html":["\n","  <div id=\"df-d57e5973-0b42-449b-901e-7d48d20dc21a\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>experiment_id</th>\n","      <th>protein_name</th>\n","      <th>pdb_id</th>\n","      <th>chain</th>\n","      <th>position</th>\n","      <th>wild_type</th>\n","      <th>mutation</th>\n","      <th>ddG</th>\n","      <th>sequence</th>\n","      <th>is_in_catalytic_pocket</th>\n","      <th>is_essential</th>\n","      <th>mt_sequence</th>\n","      <th>pdb_id_list</th>\n","      <th>pdb_id_len</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>5081</th>\n","      <td>PT008738</td>\n","      <td>U1 small nuclear ribonucleoprotein A</td>\n","      <td>1OIA</td>\n","      <td>A</td>\n","      <td>86</td>\n","      <td>Y</td>\n","      <td>F</td>\n","      <td>2.60</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>[1OIA]</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>5080</th>\n","      <td>VB01735</td>\n","      <td>U1 small nuclear ribonucleoprotein A</td>\n","      <td>1OIA</td>\n","      <td>A</td>\n","      <td>78</td>\n","      <td>Y</td>\n","      <td>F</td>\n","      <td>0.00</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>[1OIA]</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>5079</th>\n","      <td>VB04793</td>\n","      <td>U1 small nuclear ribonucleoprotein A</td>\n","      <td>1OIA</td>\n","      <td>A</td>\n","      <td>78</td>\n","      <td>Y</td>\n","      <td>D</td>\n","      <td>4.30</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>[1OIA]</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>5078</th>\n","      <td>PT008735</td>\n","      <td>U1 small nuclear ribonucleoprotein A</td>\n","      <td>1OIA</td>\n","      <td>A</td>\n","      <td>31</td>\n","      <td>Y</td>\n","      <td>S</td>\n","      <td>2.30</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLSAIFSQFGQILDILVS...</td>\n","      <td>[1OIA]</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>5077</th>\n","      <td>VB01730</td>\n","      <td>U1 small nuclear ribonucleoprotein A</td>\n","      <td>1OIA</td>\n","      <td>A</td>\n","      <td>31</td>\n","      <td>Y</td>\n","      <td>F</td>\n","      <td>0.00</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVS...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MAVPETRPNHTIYINNLNEKIKKDELKKSLFAIFSQFGQILDILVS...</td>\n","      <td>[1OIA]</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>898</th>\n","      <td>PT000404</td>\n","      <td>Dihydrofolate reductase</td>\n","      <td>1RX4|1DDR|1DDR|1DYJ|5DFR</td>\n","      <td>A</td>\n","      <td>31</td>\n","      <td>F</td>\n","      <td>V</td>\n","      <td>1.50</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWVKRNTLNKPVIMGRHT...</td>\n","      <td>[5DFR, 1DDR, 1DYJ, 1RX4]</td>\n","      <td>4</td>\n","    </tr>\n","    <tr>\n","      <th>966</th>\n","      <td>PT014402</td>\n","      <td>Dihydrofolate reductase</td>\n","      <td>1RX4|1DDR|1DDR|1DYJ|5DFR</td>\n","      <td>A</td>\n","      <td>122</td>\n","      <td>D</td>\n","      <td>A</td>\n","      <td>1.60</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>[5DFR, 1DDR, 1DYJ, 1RX4]</td>\n","      <td>4</td>\n","    </tr>\n","    <tr>\n","      <th>965</th>\n","      <td>PT002910</td>\n","      <td>Dihydrofolate reductase</td>\n","      <td>1RX4|1DDR|1DDR|1DYJ|5DFR</td>\n","      <td>A</td>\n","      <td>121</td>\n","      <td>G</td>\n","      <td>Y</td>\n","      <td>0.20</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>[5DFR, 1DDR, 1DYJ, 1RX4]</td>\n","      <td>4</td>\n","    </tr>\n","    <tr>\n","      <th>964</th>\n","      <td>VB05467</td>\n","      <td>Dihydrofolate reductase</td>\n","      <td>1RX4|1DDR|1DDR|1DYJ|5DFR</td>\n","      <td>A</td>\n","      <td>121</td>\n","      <td>G</td>\n","      <td>V</td>\n","      <td>1.10</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>[5DFR, 1DDR, 1DYJ, 1RX4]</td>\n","      <td>4</td>\n","    </tr>\n","    <tr>\n","      <th>963</th>\n","      <td>PT002903</td>\n","      <td>Dihydrofolate reductase</td>\n","      <td>1RX4|1DDR|1DDR|1DYJ|5DFR</td>\n","      <td>A</td>\n","      <td>121</td>\n","      <td>G</td>\n","      <td>S</td>\n","      <td>-0.01</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHT...</td>\n","      <td>[5DFR, 1DDR, 1DYJ, 1RX4]</td>\n","      <td>4</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5086 rows × 14 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d57e5973-0b42-449b-901e-7d48d20dc21a')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-d57e5973-0b42-449b-901e-7d48d20dc21a button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-d57e5973-0b42-449b-901e-7d48d20dc21a');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","    <div id=\"df-e43552de-ee40-41bb-9e7a-792fcf6cec43\">\n","      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-e43552de-ee40-41bb-9e7a-792fcf6cec43')\"\n","                title=\"Suggest charts\"\n","                style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","      </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","      <script>\n","        async function quickchart(key) {\n","          const quickchartButtonEl =\n","            document.querySelector('#' + key + ' button');\n","          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","          quickchartButtonEl.classList.add('colab-df-spinner');\n","          try {\n","            const charts = await google.colab.kernel.invokeFunction(\n","                'suggestCharts', [key], {});\n","          } catch (error) {\n","            console.error('Error during call to suggestCharts:', error);\n","          }\n","          quickchartButtonEl.classList.remove('colab-df-spinner');\n","          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","        }\n","        (() => {\n","          let quickchartButtonEl =\n","            document.querySelector('#df-e43552de-ee40-41bb-9e7a-792fcf6cec43 button');\n","          quickchartButtonEl.style.display =\n","            google.colab.kernel.accessAllowed ? 'block' : 'none';\n","        })();\n","      </script>\n","    </div>\n","\n","    </div>\n","  </div>\n"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"dataframe","summary":"{\n  \"name\": \"df_fire_prot_median_ddG\",\n  \"rows\": 5086,\n  \"fields\": [\n    {\n      \"column\": \"experiment_id\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5086,\n        \"samples\": [\n          \"PT012593\",\n          \"VB03007\",\n          \"VB04168\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"protein_name\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 158,\n        \"samples\": [\n          \"Thermonuclease\",\n          \"Cytochrome c isoform 1\",\n          \"Sialidase\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pdb_id\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 174,\n        \"samples\": [\n          \"1UWO|1UWO|2H61\",\n          \"1W4E|1W3D\",\n          \"1DPM|1DPM\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chain\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          \"X\",\n          \"B\",\n          \"A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"position\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 677,\n        \"min\": 1,\n        \"max\": 12762,\n        \"num_unique_values\": 534,\n        \"samples\": [\n          143,\n          137,\n          170\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"wild_type\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 20,\n        \"samples\": [\n          \"Y\",\n          \"H\",\n          \"Q\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mutation\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 20,\n        \"samples\": [\n          \"F\",\n          \"H\",\n          \"N\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"ddG\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1.8067263519453443,\n        \"min\": -13.7,\n        \"max\": 17.4,\n        \"num_unique_values\": 612,\n        \"samples\": [\n          3.38,\n          0.14,\n          -0.3\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"sequence\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 174,\n        \"samples\": [\n          \"MSELEKAMVALIDVFHQYSGREGDKHKLKKSELKELINNELSHFLEEIKEQEVVDKVMETLDNDGDGECDFQEFMAFVAMVTTACHEFFEHE\",\n          \"MAFEFKLPDIGEGIHEGEIVKWFVKPGDEVNEDDVLCEVQNDKAVVEIPSPVKGKVLEILVPEGTVATVGQTLITLDAPGYENMTFKGQEQEEAKKEEKTETVSKEEKVDAVAPNAPAAEAEAGPNRRVIAMPSVRKYAREKGVDIRLVQGTGKNGRVLKEDIDAFLAGGAKPAPAAAEEKAAPAAAKPATTEGEFPETREKMSGIRRAIAKAMVHSKHTAPHVTLMDEADVTKLVAHRKKFKAIAAEKGIKLTFLPYVVKALVSALREYPVLNTSIDDETEEIIQKHYYNIGIAADTDRGLLVPVIKHADRKPIFALAQEINELAEKARDGKLTPGEMKGASCTITNIGSAGGQWFTPVINHPEVAILGIGRIAEKPIVRDGEIVAAPMLALSLSFDHRMIDGATAQKALNHIKRLLSDPELLLMEA\",\n          \"MQTRRVVLKSAAAAGTLLGGLAGCASVAGSIGTGDRINTVRGPITISEAGFTLTHEHICGSSAGFLRAWPEFFGSRKALAEKAVRGLRRARAAGVRTIVDVSTFDIGRDVSLLAEVSRAADVHIVAATGLWFDPPLSMRLRSVEELTQFFLREIQYGIEDTGIRAGIIKVATTGKATPFQELVLKAAARASLATGVPVTTHTAASQRDGEQQAAIFESEGLSPSRVCIGHSDDTDDLSYLTALAARGYLIGLDHIPHSAIGLEDNASASALLGIRSWQTRALLIKALIDQGYMKQILVSNDWLFGFSSYVTNIMDVMDRVNPDGMAFIPLRVIPFLREKGVPQETLAGITVTNPARFLSPTLRAS\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"is_in_catalytic_pocket\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          true,\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"is_essential\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          true,\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mt_sequence\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5086,\n        \"samples\": [\n          \"MLVMTEYLLSAGICMAIVSILLIGMAISNVSKGQYAKRFFFFATSCLVLTLVVVSSLSSSANASQTDNGVNRSGSEDPTVYSATSTKKLHKEPATLIKAIDGDTVKLMYKGQPMTFRLLLVDTPETKHPKKGVEKYGPEASAFTKKMVENAKKIEIEFDKGQRTDKYGRGLAYIYADGKMVNEALVRQGLAKVAYVYKPNNTHEQHLRKSEAQAKKEKLNIWSEDNADSGQ\",\n          \"MTDITANVVVSNPRPIFTESRSFKAVANGKIYIGQIDTDPVNPANQIPVYIENEDGSHVQITQPLIINAAGKIVYNGQLVKIVTVQGHSMAIYDANGSQVDYIANVLKYDPDQYSIEADKKFKYSVKLSDYPTLQDAASAAVDGLLIDRDYNFYGGETVDFGGKVLTIECKAKFIGDGNLIFTKLGKGSRIAGVFMESTTTPWVIKPWTDDNQWLTDAAAVVATLKQSKTDGYQPTVSDYVKFPRIETLLPPNAKGQNITSTLEIRECIGVEVHRASGLMAGFLFRGCHFCKMVDANNPSGGKDGIITFENLSGDWGKGNYVIGGRTSYGSVSSAQFLRNNGGFERDGGVIGFTSYRAGESGVKTWQGTVGSTTSRNYNLQFRDSVVIYPVWDGFDLGADTDMNPELDRPGDYPITQYPLHQLPLNHLIDNLLVRGALGVGFGMDGKGMYVSNITVEDCAGSGAYLLTHESVFTNIAIIDTNTKDFQANQIYISGACRVNGLRLIGIRSTDGQGLTIDAPNSTVSGITGMVDPSRINVANLAEEGLGNIRANSFGYDSAAIKLRIHKLSKTLDSGALYSHINGGAGSGSAYTQLTAISGSTPDAVSLKVNHKDCRGAEIPFVPDIASDDFIKDSSCFLPYWENNSTSLKALVKKPNGELVRLTLATL\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pdb_id_list\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pdb_id_len\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": 1,\n        \"max\": 4,\n        \"num_unique_values\": 4,\n        \"samples\": [\n          2,\n          4\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"}},"metadata":{},"execution_count":167}]},{"cell_type":"code","source":["df_temp = df_fire_prot_median_ddG[df_fire_prot_median_ddG[\"experiment_id\"] == \"PT008738\"]\n","print(df_temp[\"sequence\"].values[0][85])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"O7kiVy-7tipF","executionInfo":{"status":"ok","timestamp":1746587719178,"user_tz":240,"elapsed":29,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"26c6e7a6-df1b-4f48-be41-96288623f0bb"},"execution_count":174,"outputs":[{"output_type":"stream","name":"stdout","text":["Y\n"]}]},{"cell_type":"code","source":["test = \"MAVPETRPNHTIYINNLNEKIKKDELKKSLYAIFSQFGQILDILVSRSLKMRGQAFVIFKEVSSATNALRSMQGFPFYDKPMRIQYAKTDSDIIA\"\n","test[85]"],"metadata":{"id":"I9HP78Jcv0Rs","executionInfo":{"status":"ok","timestamp":1746587836312,"user_tz":240,"elapsed":12,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"bb6e1a54-bca5-42bb-c3af-6c3582011ad2","colab":{"base_uri":"https://localhost:8080/","height":35}},"execution_count":175,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'Y'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":175}]},{"cell_type":"code","source":["columns_to_keep = [\n","        'experiment_id', 'protein_name', 'uniprot_id', 'pdb_id', 'chain',\n","        'position', 'wild_type', 'mutation', 'ddG', 'sequence',\n","        'is_in_catalytic_pocket', 'is_essential'\n","]\n","# df_fireprot = df_fireprot[columns_to_keep].dropna().drop_duplicates(subset=['experiment_id'])"],"metadata":{"id":"vpf7RuQxLKlj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# take shyam's code to generate the dataset\n","# make it run for 100 sequences => time it takes to run\n","# understand what is taking the longest: calling the API to get the pdbs or building the graphs everytime\n","# iterrows might be inefficient: perhaps better to group by wt sequence, then maybe parallel compute the calls to the API"],"metadata":{"id":"Ov8-qNb7Lx-8"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"AYf4-dsAnTjQ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Extract PDB file from uniprot id (wild-type)\n","\n","def fetch_alphafold_pdb(uniprot_id, save_dir=\"wt_pdbs\"):\n","    os.makedirs(save_dir, exist_ok=True)\n","    url = f\"https://alphafold.ebi.ac.uk/files/AF-{uniprot_id}-F1-model_v4.pdb\"\n","    save_path = os.path.join(save_dir, f\"{uniprot_id}.pdb\")\n","    if not os.path.exists(save_path):\n","        r = requests.get(url)\n","        if r.status_code == 200:\n","            with open(save_path, 'wb') as f:\n","                f.write(r.content)\n","        else:\n","            raise ValueError(f\"Could not fetch AlphaFold PDB for {uniprot_id}\")\n","    return save_path\n","\n","# For the mutations becuase the unique PDBs used in the database correspond to the mutant protein\n","\n","def fetch_pdb_on_demand(pdb_id, save_dir=\"pdbs\"):\n","    pdbl = PDBList()\n","    pdb_id = pdb_id.lower()\n","    os.makedirs(save_dir, exist_ok=True)\n","    filepath = pdbl.retrieve_pdb_file(pdb_id, pdir=save_dir, file_format='pdb')\n","    actual_pdb_path = os.path.join(save_dir, f\"{pdb_id}.pdb\")\n","    if not os.path.exists(actual_pdb_path):\n","        os.rename(filepath, actual_pdb_path)\n","    return actual_pdb_path\n","\n","\n","# Use the PDB to extract coordinates of residues (structural info)\n","\n","def extract_residue_coords(pdb_file, chain_id):\n","    parser = PDBParser(QUIET=True)\n","    structure = parser.get_structure('protein', pdb_file)\n","    model = structure[0]\n","    coords = {}\n","    for residue in model[chain_id]:\n","        if 'CA' in residue:\n","            res_id = residue.get_id()[1]\n","            coords[res_id] = residue['CA'].get_coord()\n","    return coords\n","\n","# build graph structure graph where nodes = residues, edges = distance (angstroms)\n","\n","def build_protein_graph(coords, distance_threshold=8.0):\n","    G = nx.Graph()\n","    residue_ids = list(coords.keys())\n","    for res_id in residue_ids:\n","        G.add_node(res_id, pos=coords[res_id])\n","    for i, id1 in enumerate(residue_ids):\n","        for id2 in residue_ids[i+1:]:\n","            dist = euclidean(coords[id1], coords[id2])\n","            if dist <= distance_threshold:\n","                G.add_edge(id1, id2, distance=dist)\n","    return G\n","\n","    # 8.0 angstrom threshold back up by literature: https://pmc.ncbi.nlm.nih.gov/articles/PMC3549380/\n","\n","# Construct the graph from predefined structure and populate with PDB extracted residue coordinates\n","\n","def construct_graph_from_pdb(pdb_file, chain_id, distance_threshold=8.0):\n","    coords = extract_residue_coords(pdb_file, chain_id)\n","    return build_protein_graph(coords, distance_threshold)\n","\n","# One hot encode the protein based on the 20 amino acids\n","\n","def one_hot_encode_aa(aa):\n","    aa_list = list(\"ACDEFGHIKLMNPQRSTVWY\")\n","    return [int(aa == x) for x in aa_list]\n","\n","# Enrich the nodes with relevant functional features (amino acid type one-hot, catalytic pocket, position, relative pos, etc)\n","\n","def enrich_graph_with_features(graph, row, sequence_length, mutate=False):\n","    sequence = row['sequence']\n","    catalytic = row['is_in_catalytic_pocket']\n","    essential = row['is_essential']\n","    mutation_pos = int(row['position'])\n","    mutant_aa = row['mutation']\n","    pdb_positions = sorted(graph.nodes)\n","    offset = min(pdb_positions)\n","\n","    for pos in graph.nodes:\n","        seq_index = pos - offset\n","        if 0 <= seq_index < len(sequence):\n","            aa = sequence[seq_index]\n","            if mutate and pos == mutation_pos:\n","                aa = mutant_aa\n","            graph.nodes[pos]['aa_type'] = one_hot_encode_aa(aa)\n","            graph.nodes[pos]['is_catalytic'] = int(catalytic)\n","            graph.nodes[pos]['is_essential'] = int(essential)\n","            graph.nodes[pos]['relative_pos'] = pos / sequence_length\n","    return graph\n","\n","# Convert the graph from nx to pytorch object for graph processing\n","\n","def convert_nx_to_pyg(graph, ddG_value=0.0):\n","    features = []\n","    node_id_map = {}\n","    filtered_nodes = []\n","    for i, n in enumerate(graph.nodes):\n","        node_data = graph.nodes[n]\n","        if all(k in node_data for k in ['aa_type', 'is_catalytic', 'is_essential', 'relative_pos']):\n","            node_id_map[n] = len(filtered_nodes)\n","            filtered_nodes.append(n)\n","            feat = node_data['aa_type'] + [\n","                node_data['is_catalytic'],\n","                node_data['is_essential'],\n","                node_data['relative_pos']\n","            ]\n","            features.append(feat)\n","    if len(features) == 0:\n","        raise ValueError(\"No usable nodes with complete features.\")\n","    edge_index, edge_attr = [], []\n","    for u, v, attrs in graph.edges(data=True):\n","        if u in node_id_map and v in node_id_map:\n","            edge_index.append([node_id_map[u], node_id_map[v]])\n","            edge_index.append([node_id_map[v], node_id_map[u]])\n","            edge_attr.append([attrs['distance']])\n","            edge_attr.append([attrs['distance']])\n","    if len(edge_index) == 0:\n","        raise ValueError(\"No valid edges in graph.\")\n","    x = torch.tensor(features, dtype=torch.float)\n","    edge_index = torch.tensor(edge_index, dtype=torch.long).t().contiguous()\n","    edge_attr = torch.tensor(edge_attr, dtype=torch.float)\n","    y = torch.tensor([ddG_value], dtype=torch.float)\n","    return Data(x=x, edge_index=edge_index, edge_attr=edge_attr, y=y)"],"metadata":{"id":"CZ-fDhm-6qrJ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Define DeltaDDGDataset class which extracts PDB structure and saves it per entry\n","\n","# Note for future: consider storing the PDB files as a folder on Dropbox (.tar) for efficiency and speed. Store all PDB structures in one place.\n","\n","\n","# Global cache (can optionally persist to disk later)\n","structure_cache = {}\n","\n","class DeltaDDGDataset(Dataset):\n","    def __init__(self, df):\n","        self.samples = []\n","\n","        for _, row in tqdm(df.iterrows(), total=len(df), desc=\"Building dataset\"):\n","            try:\n","                pdb_id = str(row['pdb_id']).split('|')[0].lower()\n","                uniprot_id = row['uniprot_id']\n","\n","                # Cache structure fetches\n","                if pdb_id in structure_cache:\n","                    pdb_mut = structure_cache[pdb_id]\n","                else:\n","                    pdb_mut = fetch_pdb_on_demand(pdb_id)\n","                    structure_cache[pdb_id] = pdb_mut\n","\n","                if uniprot_id in structure_cache:\n","                    pdb_wt = structure_cache[uniprot_id]\n","                else:\n","                    pdb_wt = fetch_alphafold_pdb(uniprot_id)\n","                    structure_cache[uniprot_id] = pdb_wt\n","\n","                # Graph construction\n","                G_mut = construct_graph_from_pdb(pdb_mut, row['chain'])\n","                G_mut = enrich_graph_with_features(G_mut, row, len(row['sequence']), mutate=True)\n","                data_mut = convert_nx_to_pyg(G_mut, row['ddG'])\n","                data_mut.batch = torch.zeros(data_mut.num_nodes, dtype=torch.long)\n","\n","                G_wt = construct_graph_from_pdb(pdb_wt, 'A')\n","                G_wt = enrich_graph_with_features(G_wt, row, len(row['sequence']), mutate=False)\n","                data_wt = convert_nx_to_pyg(G_wt, row['ddG'])\n","                data_wt.batch = torch.zeros(data_wt.num_nodes, dtype=torch.long)\n","\n","                ddG = torch.tensor(row['ddG'], dtype=torch.float)\n","                metadata = row.to_dict()\n","\n","                self.samples.append((data_wt, data_mut, ddG, metadata))\n","\n","            except Exception as e:\n","                print(f\"Skipping row due to error: {e}\")\n","\n","    def __len__(self):\n","        return len(self.samples)\n","\n","    def __getitem__(self, idx):\n","        return self.samples[idx]  # Always returns (data_wt, data_mut, ddG, metadata)\n","\n","    def save(self, path):\n","        data = {\n","            'samples': self.samples,\n","        }\n","        torch.save(data, path)\n","\n","    @classmethod\n","    def load(cls, path):\n","        data = torch.load(path, weights_only=False)\n","        obj = cls.__new__(cls)\n","        obj.samples = data['samples']\n","        return obj\n"],"metadata":{"id":"DAJqOZqp683R"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Fusion Predictor Class for both ESM and GNN\n","\n","from transformers import AutoTokenizer, EsmModel\n","\n","class DDGFusionPredictor(nn.Module):\n","    def __init__(self,\n","                 esm_model_name=\"facebook/esm2_t6_8M_UR50D\",\n","                 embedding_dim=320,\n","                 conv_channels=128,\n","                 heads=4,\n","                 ffn_dim=256,\n","                 hidden_dim = 256,\n","                 in_dim = 23):\n","        super().__init__()\n","\n","        # Device setup for trainable layers\n","        self.device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n","\n","        # Load pretrained ESM model on CPU\n","        self.tokenizer = AutoTokenizer.from_pretrained(esm_model_name)\n","        self.esm_model = EsmModel.from_pretrained(esm_model_name).eval()\n","        for param in self.esm_model.parameters():\n","            param.requires_grad = False\n","\n","        self.conv1 = GCNConv(in_dim,    hidden_dim)\n","        self.conv2 = GCNConv(hidden_dim, hidden_dim)\n","        self.conv3 = GCNConv(hidden_dim, hidden_dim)\n","        self.conv4 = GCNConv(hidden_dim, hidden_dim)\n","\n","        self.gnorm1 = GraphNorm(hidden_dim)\n","        self.gnorm2 = GraphNorm(hidden_dim)\n","        self.gnorm3 = GraphNorm(hidden_dim)\n","        self.gnorm4 = GraphNorm(hidden_dim)\n","        self.embedding_dim = self.esm_model.config.hidden_size\n","        self.project = nn.Linear(hidden_dim, self.embedding_dim)\n","\n","        # Convolutional encoder\n","        self.conv1d = nn.Conv1d(2*self.embedding_dim, conv_channels, kernel_size=3, padding=1)\n","\n","        # Multi-head attention\n","        self.attention = nn.MultiheadAttention(embed_dim=conv_channels, num_heads=heads, batch_first=True)\n","\n","        self.norm1 = nn.LayerNorm(conv_channels)\n","        self.norm2 = nn.LayerNorm(conv_channels)\n","\n","        # Position-wise feedforward\n","        self.ffn = nn.Sequential(\n","            nn.Linear(conv_channels, ffn_dim),\n","            nn.ReLU(),\n","            nn.Linear(ffn_dim, conv_channels)\n","        )\n","        self.dropout = nn.Dropout(0.3)\n","        # Pooling layers\n","        self.global_avg_pool = nn.AdaptiveAvgPool1d(1)\n","        self.global_max_pool = nn.AdaptiveMaxPool1d(1)\n","\n","        #Regression head\n","        self.regressor = nn.Sequential(\n","          nn.Linear(conv_channels * 2, 256),\n","          nn.ReLU(),\n","          nn.Dropout(0.3),\n","          nn.Linear(256, 64),\n","          nn.ReLU(),\n","          nn.Linear(64, 1)\n","       )\n","\n","\n","    def forward(self, wt_seq, mt_seq, data_wt, data_mut):\n","        # Tokenize and compute embeddings on CPU without gradients\n","        with torch.no_grad():\n","            wt_tokens = self.tokenizer(wt_seq, return_tensors=\"pt\", padding=True, truncation=True,max_length=1024)\n","            mt_tokens = self.tokenizer(mt_seq, return_tensors=\"pt\", padding=True, truncation=True,max_length=1024)\n","\n","            wt_tokens = {k: v.to(self.device) for k, v in wt_tokens.items()}\n","            mt_tokens = {k: v.to(self.device) for k, v in mt_tokens.items()}\n","\n","            wt_embed = self.esm_model(**wt_tokens).last_hidden_state[:, 1:-1, :]  # remove [CLS] and [SEP]\n","            mt_embed = self.esm_model(**mt_tokens).last_hidden_state[:, 1:-1, :]\n","\n","            d = (wt_embed - mt_embed) # Move difference to GPU\n","\n","        # Encde\n","        x0_wt = F.relu(self.gnorm1(self.conv1(data_wt.x, data_wt.edge_index)))\n","        x0_mut = F.relu(self.gnorm1(self.conv1(data_mut.x, data_mut.edge_index)))\n","\n","        x1_wt = F.relu(self.gnorm2(self.conv2(x0_wt, data_wt.edge_index))) + x0_wt\n","        x1_mut = F.relu(self.gnorm2(self.conv2(x0_mut, data_mut.edge_index))) + x0_mut\n","\n","        # x2 = F.relu(self.norm3(self.conv3(x1, data_wt.edge_index))) + x1\n","        x2_wt = F.relu(self.gnorm3(self.conv3(x1_wt, data_wt.edge_index))) + x1_wt\n","        x2_mut = F.relu(self.gnorm3(self.conv3(x1_mut, data_mut.edge_index))) + x1_mut\n","\n","        x3_wt = F.relu(self.gnorm4(self.conv4(x2_wt, data_wt.edge_index))) + x2_wt\n","        x3_mut = F.relu(self.gnorm4(self.conv4(x2_mut, data_mut.edge_index))) + x2_mut\n","\n","        h_wt = global_mean_pool(x3_wt, data_wt.batch)\n","        h_mut = global_mean_pool(x3_mut, data_mut.batch)\n","\n","        end_g_encode_wt = self.project(h_wt)\n","        end_g_encode_mut = self.project(h_mut)\n","\n","        d_g_encode = end_g_encode_wt - end_g_encode_mut\n","\n","\n","        B, L, D_esm = d.shape  # Original shape: [B, L, D_esm]\n","\n","        # Reshape d_g_encode to [B, L, D_esm]\n","        d_g_encode_exp = d_g_encode.unsqueeze(1).expand(-1, L, D_esm)\n","\n","        # conc_emb = torch.cat([d_t, d_g_encode_exp], dim=1)  # Original shape: [B, 2*D_esm, L]\n","        conc_emb = torch.cat([d, d_g_encode_exp], dim=2)  # Shape: [B, L, 2*D_esm]\n","\n","        # Conv1D (expects B, C, L) and transpose back\n","        c = self.conv1d(conc_emb.transpose(1, 2)).transpose(1, 2)  # Expected shape: (B, L, C)\n","\n","        # Multi-head attention + residual\n","        m, _ = self.attention(c, c, c)\n","\n","        z = self.norm1(c + m)\n","\n","        # Position-wise feedforward + residual\n","        f = self.norm2(z + self.dropout(self.ffn(z)))\n","\n","        # Global pooling\n","        f_t = f.transpose(1, 2)  # (B, C, L)\n","        gp = self.global_avg_pool(f_t).squeeze(-1)  # (B, C)\n","        gm = self.global_max_pool(f_t).squeeze(-1)  # (B, C)\n","        conc = torch.cat([gp, gm], dim=1)  # (B, 2C)\n","        conc = self.dropout(conc)\n","\n","        # Final regression output\n","        ddg_pred = self.regressor(conc).view(-1)\n","        return ddg_pred"],"metadata":{"id":"LM-OmbUh7T6l"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#load the data from saved files\n","\n","train_dataset_loaded = DeltaDDGDataset.load(\"/content/drive/MyDrive//BMI_707_Project/707_Files_for_Colab/official_train_set.pkl\")\n","val_dataset_loaded = DeltaDDGDataset.load(\"/content/drive/MyDrive//BMI_707_Project/707_Files_for_Colab/official_val_set.pkl\")\n","\n","#convert to dataloader object for training\n","\n","# train_loader_official = DataLoader(train_dataset_loaded, batch_size=4, shuffle=True)\n","# val_loader_official = DataLoader(val_dataset_loaded, batch_size=4)\n"],"metadata":{"id":"391wPfsh7O8m"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#align 2 sequences\n","from Bio import Align\n","from Bio.Align import PairwiseAligner\n","import time"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"qR9wpCSqDzKe","executionInfo":{"status":"ok","timestamp":1746511038396,"user_tz":240,"elapsed":11,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"62a67dc1-1970-4c89-d72d-0180b25b59c1"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["0.000522613525390625\n"]}]},{"cell_type":"code","source":["def get_alignment_from_graphs(graph, sequence):\n","  aa_feature_matrix = graph.x[:,:20]\n","  aa_np = aa_feature_matrix.numpy()\n","  aa_list_from_np = [idx2aa[idx] for idx in np.argmax(aa_np, axis=1)]\n","  aa_seq_from_np = \"\".join(aa_list_from_np)\n","  aligner = PairwiseAligner()\n","  aligner.mode = 'global'\n","  aligner.gap_score = -1\n","  aligner.target_gap_score = -float('inf')\n","  alignments = aligner.align(sequence,aa_seq_from_np)\n","  return alignments[0].indices[1,:]\n","\n","# t = time.time()\n","for i in tqdm(range(len(train_dataset_loaded))):\n","  temp = train_dataset_loaded[i]\n","  wt_graph = temp[0]\n","  mut_graph = temp[1]\n","  metadata = temp[3]\n","  wt_seq = metadata[\"sequence\"]\n","  mut_seq = list(wt_seq)\n","  mut_seq[metadata[\"position\"]-1] = metadata[\"mutation\"]\n","  mut_seq = \"\".join(mut_seq)\n","  wt_alignment = get_alignment_from_graphs(wt_graph, wt_seq)\n","  mut_alignment = get_alignment_from_graphs(mut_graph, mut_seq)\n","# print((time.time()-t)/100)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"KDweCqcvUx9a","executionInfo":{"status":"ok","timestamp":1746514303263,"user_tz":240,"elapsed":29009,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"3346a13a-678d-41b9-acbf-1d7c30444b00"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["100%|██████████| 8221/8221 [00:28<00:00, 283.56it/s]\n"]}]},{"cell_type":"code","source":["train_dataset_loaded[-1]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"9Q1xPcSSXYj3","executionInfo":{"status":"ok","timestamp":1746514376176,"user_tz":240,"elapsed":20,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"69c86f1e-9908-4c6c-b435-877fdffce2ee"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(Data(x=[231, 23], edge_index=[2, 2004], edge_attr=[2004, 1], y=[1], batch=[231]),\n"," Data(x=[136, 23], edge_index=[2, 1334], edge_attr=[1334, 1], y=[1], batch=[136]),\n"," tensor(2.6000),\n"," {'experiment_id': 'VB05124',\n","  'protein_name': 'Thermonuclease',\n","  'uniprot_id': 'P00644',\n","  'pdb_id': '1STN|1EY0',\n","  'chain': 'A',\n","  'position': 183,\n","  'wild_type': 'E',\n","  'mutation': 'G',\n","  'ddG': 2.6,\n","  'sequence': 'MLVMTEYLLSAGICMAIVSILLIGMAISNVSKGQYAKRFFFFATSCLVLTLVVVSSLSSSANASQTDNGVNRSGSEDPTVYSATSTKKLHKEPATLIKAIDGDTVKLMYKGQPMTFRLLLVDTPETKHPKKGVEKYGPEASAFTKKMVENAKKIEVEFDKGQRTDKYGRGLAYIYADGKMVNEALVRQGLAKVAYVYKPNNTHEQHLRKSEAQAKKEKLNIWSEDNADSGQ',\n","  'is_in_catalytic_pocket': False,\n","  'is_essential': False})"]},"metadata":{},"execution_count":183}]},{"cell_type":"code","source":["train_dataset_loaded[0]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"v_MjvxSCXpYr","executionInfo":{"status":"ok","timestamp":1746514387729,"user_tz":240,"elapsed":20,"user":{"displayName":"Valentin Badea","userId":"01447547954341807379"}},"outputId":"6431dc90-b749-4cc3-879e-eb9b360e6344"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(Data(x=[231, 23], edge_index=[2, 2004], edge_attr=[2004, 1], y=[1], batch=[231]),\n"," Data(x=[136, 23], edge_index=[2, 1334], edge_attr=[1334, 1], y=[1], batch=[136]),\n"," tensor(0.6000),\n"," {'experiment_id': 'VB05069',\n","  'protein_name': 'Thermonuclease',\n","  'uniprot_id': 'P00644',\n","  'pdb_id': '1STN|1EY0',\n","  'chain': 'A',\n","  'position': 164,\n","  'wild_type': 'T',\n","  'mutation': 'S',\n","  'ddG': 0.6,\n","  'sequence': 'MLVMTEYLLSAGICMAIVSILLIGMAISNVSKGQYAKRFFFFATSCLVLTLVVVSSLSSSANASQTDNGVNRSGSEDPTVYSATSTKKLHKEPATLIKAIDGDTVKLMYKGQPMTFRLLLVDTPETKHPKKGVEKYGPEASAFTKKMVENAKKIEVEFDKGQRTDKYGRGLAYIYADGKMVNEALVRQGLAKVAYVYKPNNTHEQHLRKSEAQAKKEKLNIWSEDNADSGQ',\n","  'is_in_catalytic_pocket': True,\n","  'is_essential': False})"]},"metadata":{},"execution_count":184}]}]}